app.ts

```diff
@@ -0,0 +1,211 @@
+/** @format */
+
+interface IUser {
+    firstName: string;
+    secondName: string;
+    role: string;
+    age: number;
+    gender: string;
+    isAdmin?: boolean;
+    specialization?: string;
+    course?: number;
+    faculty?: string;
+}
+
+class DataBase implements DataAggregator {
+    private static instance: DataBase;
+    private users: IUser = [];
+
+    public static getInstance(): DataBase {
+        if (!DataBase.instance) {
+            DataBase.instance = new DataBase();
+        }
+
+        return DataBase.instance;
+    }
+
+    public addUser(user: IUser): void {
+        this.users.push(user);
+    }
+
+    public getUsers(): IUser {
+        return this.users;
+    }
+
+    public removeUser(value: string, key: string): IUser {
+        return this.users.filter((item: IUser) => item[key] !== value)
+    }
+
+    public createIterator(): DataBaseIterator {
+        return new DataBaseIterator(this)
+    }
+}
+
+interface DataAggregator {
+    getUsers(): object;
+}
+
+class DataBaseIterator {
+    private position: number = 0;
+
+    constructor(private users: DataAggregator) { }
+
+    public current(index: number = this.position) {
+        this.position = index;
+        return this.users.getUsers()[index];
+    }
+
+    public dataLenght() {
+        return this.users.getUsers()
+    }
+
+    public next(): void {
+        this.position++;
+    }
+
+    public prev(): void {
+        this.position--;
+    }
+}
+
+class UserBuilder {
+
+    firstName: string;
+    secondName: string;
+    age: number;
+    gender: string;
+    protected role: string
+
+    constructor(role: string) {
+        this.role = role
+    }
+
+    public setFirstName(value: string) {
+        return this.firstName = value;
+    }
+
+    public setSecondName(value: string) {
+        return this.secondName = value;
+    }
+
+    public setAge(value: number) {
+        return this.age = value;
+    }
+
+    public setGender(value: string) {
+        return this.gender = value;
+    }
+
+    public build() {
+        return new User(this)
+    }
+
+    public getUser() {
+        return this;
+    }
+
+}
+
+class User {
+
+    firstName: string;
+    secondName: string;
+    age: number;
+    gender: string;
+
+    constructor(build: UserBuilder) {
+        this.firstName = build.firstName;
+        this.secondName = build.secondName;
+        this.age = build.age;
+        this.gender = build.gender;
+    }
+}
+
+class Admin extends UserBuilder {
```

Comment:
The name of the class is a little confusing, as it extends UserBuilder, its better to call it AdminBuilder or AdminUserBuilder. The same goes for other builders as well